File: packages/provide-uterm-go/render/palette.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package render provides ANSI palette quantizers, SGR emitters, and
// structured text segments for non-terminal clients. Port of
// provide.uterm.render (palette.py, sgr.py, segments.py).
package render

import "sync"

// RGB is a 24-bit color triple.
type RGB struct {
	R, G, B int
}

// ANSI16Entry is one standard 16-color palette entry: its RGB plus the SGR
// foreground and background codes.
type ANSI16Entry struct {
	RGB    RGB
	FGCode int
	BGCode int
}

// ANSI16Palette is the standard 16-color ANSI palette.
var ANSI16Palette = []ANSI16Entry{
	{RGB{0, 0, 0}, 30, 40},
	{RGB{170, 0, 0}, 31, 41},
	{RGB{0, 170, 0}, 32, 42},
	{RGB{170, 85, 0}, 33, 43},
	{RGB{0, 0, 170}, 34, 44},
	{RGB{170, 0, 170}, 35, 45},
	{RGB{0, 170, 170}, 36, 46},
	{RGB{170, 170, 170}, 37, 47},
	{RGB{85, 85, 85}, 90, 100},
	{RGB{255, 85, 85}, 91, 101},
	{RGB{85, 255, 85}, 92, 102},
	{RGB{255, 255, 85}, 93, 103},
	{RGB{85, 85, 255}, 94, 104},
	{RGB{255, 85, 255}, 95, 105},
	{RGB{85, 255, 255}, 96, 106},
	{RGB{255, 255, 255}, 97, 107},
}

var (
	xterm256     []RGB
	xterm256Once sync.Once
)

// buildXterm256 lazily builds the xterm 256-color palette: the standard 16,
// the 216-color cube (16-231), and 24 grays (232-255).
func buildXterm256() {
	xterm256Once.Do(func() {
		xterm256 = make([]RGB, 0, 256)
		for _, e := range ANSI16Palette {
			xterm256 = append(xterm256, e.RGB)
		}
		for ri := range 6 {
			for gi := range 6 {
				for bi := range 6 {
					level := func(i int) int {
						if i == 0 {
							return 0
						}
						return 55 + 40*i
					}
					xterm256 = append(xterm256, RGB{level(ri), level(gi), level(bi)})
				}
			}
		}
		for i := range 24 {
			v := 8 + 10*i
			xterm256 = append(xterm256, RGB{v, v, v})
		}
	})
}

func colorDistSq(a, b RGB) int {
	dr, dg, db := a.R-b.R, a.G-b.G, a.B-b.B
	return dr*dr + dg*dg + db*db
}

// Nearest16 returns the (fgCode, bgCode) pair for the nearest 16-color
// match.
func Nearest16(r, g, b int) (fgCode, bgCode int) {
	target := RGB{r, g, b}
	bestI := 0
	bestD := colorDistSq(target, ANSI16Palette[0].RGB)
	for i := 1; i < 16; i++ {
		if d := colorDistSq(target, ANSI16Palette[i].RGB); d < bestD {
			bestD = d
			bestI = i
		}
	}
	return ANSI16Palette[bestI].FGCode, ANSI16Palette[bestI].BGCode
}

// Nearest256 returns the xterm 256-color index for the nearest match.
func Nearest256(r, g, b int) int {
	buildXterm256()
	target := RGB{r, g, b}
	bestI := 0
	bestD := colorDistSq(target, xterm256[0])
	for i := 1; i < 256; i++ {
		if d := colorDistSq(target, xterm256[i]); d < bestD {
			bestD = d
			bestI = i
		}
	}
	return bestI
}

File: packages/provide-uterm-ts/src/render/palette.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * ANSI colour palettes, and mapping an arbitrary colour onto one.
 *
 * Port of `provide.uterm.render.palette`.
 *
 * Distance is squared Euclidean in RGB. Not perceptual — a perceptual metric
 * would be better and would also disagree with every other implementation of
 * this palette, which is the one thing a quantiser cannot afford.
 */

/** One standard ANSI colour: its RGB, and the codes that select it. */
export interface Ansi16Entry {
  r: number;
  g: number;
  b: number;
  /** The SGR code that sets it as a foreground. */
  fg: number;
  /** The SGR code that sets it as a background. */
  bg: number;
}

/** The sixteen standard ANSI colours, in their canonical order. */
export const ANSI16_PALETTE: readonly Ansi16Entry[] = [
  { r: 0, g: 0, b: 0, fg: 30, bg: 40 },
  { r: 170, g: 0, b: 0, fg: 31, bg: 41 },
  { r: 0, g: 170, b: 0, fg: 32, bg: 42 },
  { r: 170, g: 85, b: 0, fg: 33, bg: 43 },
  { r: 0, g: 0, b: 170, fg: 34, bg: 44 },
  { r: 170, g: 0, b: 170, fg: 35, bg: 45 },
  { r: 0, g: 170, b: 170, fg: 36, bg: 46 },
  { r: 170, g: 170, b: 170, fg: 37, bg: 47 },
  { r: 85, g: 85, b: 85, fg: 90, bg: 100 },
  { r: 255, g: 85, b: 85, fg: 91, bg: 101 },
  { r: 85, g: 255, b: 85, fg: 92, bg: 102 },
  { r: 255, g: 255, b: 85, fg: 93, bg: 103 },
  { r: 85, g: 85, b: 255, fg: 94, bg: 104 },
  { r: 255, g: 85, b: 255, fg: 95, bg: 105 },
  { r: 85, g: 255, b: 255, fg: 96, bg: 106 },
  { r: 255, g: 255, b: 255, fg: 97, bg: 107 },
];

/** An RGB triple. */
export type Rgb = readonly [number, number, number];

/** How many levels each channel of the colour cube has. */
const CUBE_LEVELS = 6;

/** The first cube step is wider than the rest, which is the easy thing to get wrong. */
const CUBE_FIRST_STEP = 55;
const CUBE_STEP = 40;

/** How many greys follow the cube, and where they start and step. */
const GREY_COUNT = 24;
const GREY_START = 8;
const GREY_STEP = 10;

/** One channel's value at a cube level. */
function cubeChannel(level: number): number {
  return level === 0 ? 0 : CUBE_FIRST_STEP + CUBE_STEP * level;
}

/** Build the xterm 256-colour palette: the standard sixteen, a cube, then greys. */
function buildXterm256(): Rgb[] {
  const palette: Rgb[] = ANSI16_PALETTE.map((entry) => [entry.r, entry.g, entry.b] as Rgb);
  for (let ri = 0; ri < CUBE_LEVELS; ri += 1) {
    for (let gi = 0; gi < CUBE_LEVELS; gi += 1) {
      for (let bi = 0; bi < CUBE_LEVELS; bi += 1) {
        palette.push([cubeChannel(ri), cubeChannel(gi), cubeChannel(bi)]);
      }
    }
  }
  for (let index = 0; index < GREY_COUNT; index += 1) {
    const value = GREY_START + GREY_STEP * index;
    palette.push([value, value, value]);
  }
  return palette;
}

/** The xterm 256-colour palette, built once. */
export const XTERM256_PALETTE: readonly Rgb[] = buildXterm256();

/** Squared Euclidean distance between two colours. */
function distanceSquared(r: number, g: number, b: number, other: Rgb): number {
  return (r - other[0]) ** 2 + (g - other[1]) ** 2 + (b - other[2]) ** 2;
}

/**
 * The index of the nearest colour in a palette.
 *
 * The comparison is strict, so the *first* nearest wins: a colour exactly
 * between two entries takes the earlier one. Every renderer that quantises has
 * to agree on that, or the same screen renders differently in two places — and
 * the xterm palette repeats two colours the standard sixteen already have, so
 * the rule is reachable rather than theoretical.
 */
function nearestIndex(r: number, g: number, b: number, palette: readonly Rgb[]): number {
  let bestIndex = 0;
  let bestDistance = distanceSquared(r, g, b, palette[0] as Rgb);
  for (let index = 1; index < palette.length; index += 1) {
    const distance = distanceSquared(r, g, b, palette[index] as Rgb);
    if (distance < bestDistance) {
      bestDistance = distance;
      bestIndex = index;
    }
  }
  return bestIndex;
}

/** The foreground and background codes for the nearest standard ANSI colour. */
export function nearest16(r: number, g: number, b: number): readonly [number, number] {
  const entry = ANSI16_PALETTE[
    nearestIndex(
      r,
      g,
      b,
      ANSI16_PALETTE.map((candidate) => [candidate.r, candidate.g, candidate.b] as Rgb),
    )
  ] as Ansi16Entry;
  return [entry.fg, entry.bg];
}

/** The xterm 256-colour index of the nearest colour. */
export function nearest256(r: number, g: number, b: number): number {
  return nearestIndex(r, g, b, XTERM256_PALETTE);
}
